File: ios/Harakat Lens/Services/Vocalizer/Vocalizer.swift
//
//  Vocalizer.swift
//  Harakat Lens
//

import Foundation

struct Vocalizer {
    private let dictionary: VocalizationDictionary

    init(dictionary: VocalizationDictionary = .shared) {
        self.dictionary = dictionary
    }

    /// If `word` is bare (no harakat), return its dictionary vocalization or
    /// the original word on miss. If `word` already contains any harakat
    /// character, return it unchanged.
    func vocalize(_ word: String) -> String {
        if Self.containsHarakat(word) { return word }
        return dictionary.lookup(word) ?? word
    }

    private static func containsHarakat(_ word: String) -> Bool {
        for scalar in word.unicodeScalars {
            let v = scalar.value
            if (0x064B ... 0x0652).contains(v) || v == 0x0670 { return true }
        }
        return false
    }
}

File: android/app/src/main/java/dev/veeso/harakatlens/arabic/Vocalizer.kt
package dev.veeso.harakatlens.arabic

/**
 * Ported verbatim from the Harakat-Lens iOS app (`Vocalizer/Vocalizer.swift`)
 * and the already-complete iOS `ArabicExample/Vocalizer`.
 */
class Vocalizer(private val dictionary: VocalizationDictionary) {

    /**
     * If [word] is bare (no harakat), return its dictionary vocalization or
     * the original word on miss. If [word] already contains any harakat
     * character, return it unchanged.
     */
    fun vocalize(word: String): String {
        if (containsHarakat(word)) return word
        return dictionary.lookup(word) ?: word
    }

    private fun containsHarakat(word: String): Boolean {
        var index = 0
        while (index < word.length) {
            val cp = word.codePointAt(index)
            index += Character.charCount(cp)
            if (cp in 0x064B..0x0652 || cp == 0x0670) return true
        }
        return false
    }
}
